dev-build/conf/routes

```diff
@@ -223,7 +223,7 @@ POST           /press/r2/batchupload
 
 GET            /admin/football                                                                                                   controllers.admin.SiteController.index
 GET            /admin/football/browse                                                                                            controllers.admin.PaBrowserController.browse
-POST           /admin/football/browserRedirect                                                                                   controllers.admin.PaBrowserController.browserSubstitution
+POST           /admin/football/browserRedirect                                                                                   controllers.admin.PaBrowserController.browserSubstitution()
 GET            /admin/football/browser/*query                                                                                    controllers.admin.PaBrowserController.browser(query)
 GET            /admin/football/player                                                                                            controllers.admin.PlayerController.playerIndex
 POST           /admin/football/player/card                                                                                       controllers.admin.PlayerController.redirectToCard
```

Comment:
I'm struggling a bit to understand why the brackets don't appear to be needed on _all_ of the controllers?

Reply:
Great question! I had to look this up and learned something new! Sorry for the essay but here is my understanding:

There are two types of zero arguments functions:

* `def foo()` (also called "empty-paren", e.g.: https://github.com/guardian/frontend/blob/main/admin/app/football/controllers/PaBrowserController.scala#L20) --> In Scala 2.12 we were able to call these functions without the parentheses but in Scala 2.13 we get a warning [causing an error](https://github.com/guardian/frontend/blob/main/project/ProjectSettings.scala#L30). 

```
[error] /Users/Ioanna_Kokkini/Projects/frontend/dev-build/conf/routes:226:1: Auto-application to `()` is deprecated. Supply the empty argument list `()` explicitly to invoke method browserSubstitution,
[error] or remove the empty argument list from its definition (Java-defined methods are exempt).
[error] In Scala 3, an unapplied method like this will be eta-expanded into a function.
[error] POST           /admin/football/browserRedirect                                                                                   controllers.admin.PaBrowserController.browserSubstitution
```

This is what the current PR and https://github.com/guardian/frontend/pull/25362 are addressing.

* `def foo` ("parameterless", e.g. https://github.com/guardian/frontend/blob/main/admin/app/football/controllers/PaBrowserController.scala#L41)

This can only be used when it "has no side-effects (purely functional). In other words, it would be acceptable to omit parentheses when calling `queue.size`, but not when calling `println()`. This convention mirrors the method declaration convention given above." ([from the docs](https://docs.scala-lang.org/style/method-invocation.html#))

As I was reading this I wondered does that mean all the parameterless functions in the codebase are "purely functional"?  And what if we turn `controllers.admin.PaBrowserController.browserSubstitution()` (line 226) to a parameterless function? I tried it and it still compiles so the compiler cannot actually check whether there are side-effects or not and this is a convention as per the docs. 

Here is another article I found which I think explains the difference well: https://softwareengineering.stackexchange.com/questions/235925/parameterless-empty-paren-methods-in-scala

Reply:
This is a great explanation - than you!